src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2412,101 +2414,110 @@ public override BoundNode VisitConditionalOperator(BoundConditionalOperator node
+                    }
+                    else if (convertedConsequenceResult.HasNullType)
+                    {
+                        Debug.Assert(!convertedAlternativeResult.HasNullType);
+                        rValueResult = convertedAlternativeResult;
+                    }
+                    else
+                    {
+                        rValueResult = new TypeWithState(resultType, convertedConsequenceResult.State.Join(convertedAlternativeResult.State));
+                    }
+
+                    lValueResult = rValueResult.ToTypeWithAnnotations();
                 }
             }
             else
             {
-                NullableAnnotation resultNullableAnnotation;
-
-                if (isConstantTrue)
+                NullableFlowState resultNullableState = (isConstantTrue, isConstantFalse) switch
                 {
-                    resultNullableAnnotation = getNullableAnnotation(consequence, consequenceResult);
-                }
-                else if (isConstantFalse)
-                {
-                    resultNullableAnnotation = getNullableAnnotation(alternative, alternativeResult);
-                }
-                else
-                {
-                    resultNullableAnnotation = getNullableAnnotation(consequence, consequenceResult).Join(getNullableAnnotation(alternative, alternativeResult));
-                }
+                    (true, _) => getNullableState(consequence, consequenceRValue),
+                    (_, true) => getNullableState(alternative, alternativeRValue),
+                    _ => getNullableState(consequence, consequenceRValue).Join(getNullableState(alternative, alternativeRValue))
+                };
 
-                visitResult = TypeWithAnnotations.Create(node.Type.SetUnknownNullabilityForReferenceTypes(), resultNullableAnnotation);
+                rValueResult = new TypeWithState(node.Type.SetUnknownNullabilityForReferenceTypes(), resultNullableState);
+                lValueResult = rValueResult.ToTypeWithAnnotations();
             }
-            LvalueResultType = visitResult;
+
+            SetResult(rValueResult, lValueResult);
 
             return null;
 
-            NullableAnnotation getNullableAnnotation(BoundExpression expr, TypeWithAnnotations type)
+            NullableFlowState getNullableState(BoundExpression expr, TypeWithState type)
             {
-                if (type.HasType)
+                if (!type.HasNullType)
                 {
-                    return type.GetValueNullableAnnotation();
+                    return type.State;
                 }
                 if (expr.IsLiteralNullOrDefault())
                 {
-                    return NullableAnnotation.Annotated;
+                    return NullableFlowState.MaybeNull;
                 }
-                return NullableAnnotation.Oblivious;
+                return NullableFlowState.NotNull;
             }
```

Comment:
This whole thing should be replaceable by type.State.  If a branch is unreachable, we should get NotNull, even for the null literal.